claat/serve.go

```diff
@@ -0,0 +1,221 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+var deps = map[string]string{
+	"webcomponentsjs":     "webcomponents/webcomponentsjs",
+	"codelab-components":  "googlecodelabs/codelab-components#^1.0.0",
+	"google-youtube":      "GoogleWebComponents/google-youtube#^1.2.1",
+	"iron-flex-layout":    "PolymerElements/iron-flex-layout#^1.0.3",
+	"iron-pages":          "PolymerElements/iron-pages#^1.0.4",
+	"iron-selector":       "PolymerElements/iron-selector#^1.0.8",
+	"iron-ajax":           "PolymerElements/iron-ajax#^1.0.8",
+	"paper-button":        "PolymerElements/paper-button#^1.0.3",
+	"paper-drawer-panel":  "PolymerElements/paper-drawer-panel#^1.0.3",
+	"paper-header-panel":  "PolymerElements/paper-header-panel#^1.0.4",
+	"paper-icon-button":   "PolymerElements/paper-icon-button#^1.0.3",
+	"paper-input":         "PolymerElements/paper-input#^1.1.1",
+	"paper-item":          "PolymerElements/paper-item#^1.0.2",
+	"paper-listbox":       "PolymerElements/paper-listbox#^1.0.0",
+	"paper-menu-button":   "PolymerElements/paper-menu-button#^1.0.3",
+	"paper-dropdown-menu": "PolymerElements/paper-dropdown-menu#^1.0.4",
+	"paper-styles":        "PolymerElements/paper-styles#^1.0.11",
+	"paper-tabs":          "PolymerElements/paper-tabs#^1.1.0",
+	"paper-toast":         "PolymerElements/paper-toast#^1.0.0",
+	"paper-toolbar":       "PolymerElements/paper-toolbar#^1.0.4",
+	"platinum-sw":         "PolymerElements/platinum-sw#^1.2.0",
+	"polymer":             "Polymer/polymer#standard-layer",
+	"url-search-params":   "WebReflection/url-search-params#^0.5.0",
+}
+
+var depsDir = "bower_components"
+var elemDir = "elements"
+var elemFile = elemDir + "/codelab.html"
+
+const codelabElem = `
+<link rel="import" href="../bower_components/google-codelab-elements/google-codelab-elements.html">
+<link rel="import" href="../bower_components/google-youtube/google-youtube.html">
+<link rel="import" href="../bower_components/iron-ajax/iron-ajax.html">
+`
+
+// cmdServe is the "claat serve ..." subcommand.
+func cmdServe() {
+	err := mkdirIfNec(depsDir)
+	if err != nil {
+		fmt.Println(err)
+		return
+	}
+	// If we don't have the required dependencies, go get 'em.
+	for k, v := range deps {
+		if _, err := os.Stat(depsDir + "/" + k); os.IsNotExist(err) {
+			s := strings.Split(v, "#")
+			url := "https://github.com/" + s[0] + "/archive/master.zip"
```

Comment:
Note that you're always downloading the latest version here, but we don't always want that.
For instance, if you have `googlecodelabs/codelab-components#^1.0.0` in the deps, then you should fetch this URL:

    github.com/googlecodelabs/codelab-components/archive/1.0.0/master.zip

The version spec a bit more flexible in bower, but I would stick with just this, using it directly in the URL path.

Reply:
ah, that's the magic I was missing, thanks. With that I think I'll get the right version of all the deps and this should start working properly. I was really close but not quite there due to version mismatches.

Reply:
I gave you incorrect URL. It should be of `https://github.com/user/repo/archive/<version>.zip` form.

Reply:
got it, thanks!